File: amalgamate.py
#!/usr/bin/env python3
"""
amalgamate.py — OpenCSTL single-header amalgamator
Usage: python amalgamate.py [--src-dir ./opencstl] [--out ./opencstl.h] [-c|--compact]

Rules:
  - #pragma once and include guards (#ifndef / #if !defined + #define … #endif)
    are stripped from each individual file and replaced with a single outer guard.
  - #include <s> at depth-0 (unconditional) are deduplicated and hoisted to
    the top of the output.
  - #include <s> inside conditional blocks (#if / #ifdef / …) are left in
    place to preserve platform guards (e.g. Windows.h under #ifdef _WIN32).
  - #include "local" are inlined recursively; already-visited files emit a stub.
  - --compact strips all C/C++ comments and blank lines for minimum size.
"""

import re
import sys
import argparse
from pathlib import Path
import shutil

# ──────────────────────────────────────────────────────────────────────────────
# Regex helpers
# ──────────────────────────────────────────────────────────────────────────────

RE_LOCAL_INCLUDE = re.compile(r'^\s*#\s*include\s*"([^"]+)"')
RE_SYSTEM_INCLUDE = re.compile(r'^\s*#\s*include\s*<([^>]+)>')
RE_PRAGMA_ONCE = re.compile(r'^\s*#\s*pragma\s+once\s*$')

RE_GUARD_IFNDEF = re.compile(r'^\s*#\s*ifndef\s+(\w+)\s*$')
RE_GUARD_IF_DEF = re.compile(r'^\s*#\s*if\s+!defined\s*\(\s*(\w+)\s*\)\s*$')
RE_GUARD_DEFINE = re.compile(r'^\s*#\s*define\s+(\w+)\s*$')

RE_IF_OPEN = re.compile(r'^\s*#\s*(if|ifdef|ifndef)\b')
RE_IF_CLOSE = re.compile(r'^\s*#\s*endif\b')


# ──────────────────────────────────────────────────────────────────────────────
# Include-guard stripping
# ──────────────────────────────────────────────────────────────────────────────

def strip_include_guard(lines):
    """Remove outermost include guard and #pragma once lines."""
    lines = [l for l in lines if not RE_PRAGMA_ONCE.match(l)]
    n = len(lines)

    idx = 0
    while idx < n and lines[idx].strip() == '':
        idx += 1
    if idx >= n:
        return lines

    m = RE_GUARD_IFNDEF.match(lines[idx]) or RE_GUARD_IF_DEF.match(lines[idx])
    if not m:
        return lines

    candidate_guard = m.group(1)
    guard_open_idx = idx

    next_idx = idx + 1
    while next_idx < n and lines[next_idx].strip() == '':
        next_idx += 1
    if next_idx >= n:
        return lines

    m2 = RE_GUARD_DEFINE.match(lines[next_idx])
    if not m2 or m2.group(1) != candidate_guard:
        return lines

    guard_define_idx = next_idx

    depth = 0
    closing_idx = None
    for j in range(n - 1, guard_open_idx, -1):
        if RE_IF_CLOSE.match(lines[j]):
            if depth == 0:
                closing_idx = j
                break
            depth -= 1
        elif RE_IF_OPEN.match(lines[j]):
            depth += 1

    if closing_idx is None:
        return lines

    skip = {guard_open_idx, guard_define_idx, closing_idx}
    return [line for j, line in enumerate(lines) if j not in skip]


# ──────────────────────────────────────────────────────────────────────────────
# C/C++ comment stripping (for --compact mode)
# ──────────────────────────────────────────────────────────────────────────────

def strip_c_comments(source):
    """Strip // and /* */ comments, preserving string and character literals."""
    out = []
    i = 0
    n = len(source)
    while i < n:
        c = source[i]
        nxt = source[i + 1] if i + 1 < n else ''

        # // line comment — skip to end of line (keep the newline)
        if c == '/' and nxt == '/':
            while i < n and source[i] != '\n':
                i += 1
            continue

        # /* block comment */
        if c == '/' and nxt == '*':
            i += 2
            while i < n - 1 and not (source[i] == '*' and source[i + 1] == '/'):
                i += 1
            i += 2
            continue

        # "string literal"
        if c == '"':
            out.append(c)
            i += 1
            while i < n and source[i] != '"':
                if source[i] == '\\' and i + 1 < n:
                    out.append(source[i])
                    out.append(source[i + 1])
                    i += 2
                else:
                    out.append(source[i])
                    i += 1
            if i < n:
                out.append(source[i])
                i += 1
            continue

        # 'character literal'
        if c == "'":
            out.append(c)
            i += 1
            while i < n and source[i] != "'":
                if source[i] == '\\' and i + 1 < n:
                    out.append(source[i])
                    out.append(source[i + 1])
                    i += 2
                else:
                    out.append(source[i])
                    i += 1
            if i < n:
                out.append(source[i])
                i += 1
            continue

        out.append(c)
        i += 1

    return ''.join(out)


def compact_source(source):
    """Strip comments, blank lines, and trailing whitespace."""
    stripped = strip_c_comments(source)
    kept = []
    for line in stripped.splitlines():
        s = line.rstrip()
        if s.strip() == '':
            continue
        kept.append(s)
    return '\n'.join(kept) + '\n'


# ──────────────────────────────────────────────────────────────────────────────
# Core recursive processor
# ──────────────────────────────────────────────────────────────────────────────

class Amalgamator:
    def __init__(self, src_dir):
        self.src_dir = src_dir
        self.visited = set()  # basenames already inlined
        self.top_sys_includes = []  # unconditional system includes (hoisted)
        self._top_sys_seen = set()

    def _hoist_system_include(self, line):
        m = RE_SYSTEM_INCLUDE.match(line)
        if m:
            key = m.group(1)
            if key not in self._top_sys_seen:
                self._top_sys_seen.add(key)
                self.top_sys_includes.append(line.rstrip())

    def process_file(self, filename, depth=0):
        """Recursively inline filename, return list of output lines."""
        path = self.src_dir / filename
        canonical = path.name

        if canonical in self.visited:
            return [f'// [already included: {filename}]\n']
        self.visited.add(canonical)

        if not path.exists():
            return [f'// [WARNING: file not found: {filename}]\n']

        raw = path.read_text(encoding='utf-8').splitlines(keepends=True)
        stripped = strip_include_guard(raw)

        sep = '=' * 78
        out = []
        out.append(f'\n// {sep}\n')
        out.append(f'// BEGIN  {filename:<30} (depth {depth})\n')
        out.append(f'// {sep}\n\n')

        cond_depth = 0  # nesting depth of conditional blocks

        for line in stripped:
            # track conditional nesting BEFORE processing the line
            if RE_IF_OPEN.match(line):
                cond_depth += 1
            elif RE_IF_CLOSE.match(line):
                cond_depth -= 1

            m_sys = RE_SYSTEM_INCLUDE.match(line)
            if m_sys:
                if cond_depth == 0:
                    # unconditional → hoist to top, don't emit in place
                    self._hoist_system_include(line)
                else:
                    # inside #ifdef / #if → keep in place (e.g. Windows.h)
                    out.append(line)
                continue

            m_local = RE_LOCAL_INCLUDE.match(line)
            if m_local:
                dep = m_local.group(1)
                if (self.src_dir / dep).exists():
                    out.extend(self.process_file(dep, depth + 1))
                else:
                    out.append('// [external, kept as-is]\n')
                    out.append(line)
                continue

            out.append(line)

        out.append(f'\n// {sep}\n')
        out.append(f'// END    {filename}\n')
        out.append(f'// {sep}\n')

        return out


# ──────────────────────────────────────────────────────────────────────────────
# Output assembly
# ──────────────────────────────────────────────────────────────────────────────

BANNER = """\
//
// ============================================================================
//  OpenCSTL — Single-Header Amalgamation
//  Auto-generated by amalgamate.py  (do NOT edit manually)
//
//  Copyright (C) 2018-2026, Kim Bomm, all rights reserved.
//  Licensed under the OpenCSTL License Agreement.
// ============================================================================
//
"""

BANNER_COMPACT = """\
// OpenCSTL — Single-Header Amalgamation (compact)
// Copyright (C) 2018-2026, Kim Bomm. Licensed under the OpenCSTL License Agreement.
"""

GUARD = '_OPENCSTL_AMALGAMATED_H'


def write_output(out_path, amalg, body_lines, compact=False):
    # Assemble body into a single string, then optionally compact.
    body_src = ''.join(body_lines)
    if compact:
        body_src = compact_source(body_src)

    with out_path.open('w', encoding='utf-8', newline='\n') as f:
        f.write(BANNER_COMPACT if compact else BANNER)
        f.write('\n')
        f.write('#pragma once\n')
        f.write(f'#ifndef {GUARD}\n')
        f.write(f'#define {GUARD}\n')
        if not compact:
            f.write('\n')

        if amalg.top_sys_includes:
            if not compact:
                f.write('// ── System includes — unconditional, deduplicated ─────────────────────\n')
            for inc in amalg.top_sys_includes:
                f.write(inc + '\n')
            if not compact:
                f.write('\n')

        f.write(body_src)
        if not body_src.endswith('\n'):
            f.write('\n')

        if not compact:
            f.write('\n')
        f.write(f'#endif // {GUARD}\n')


# ──────────────────────────────────────────────────────────────────────────────
# Entry point
# ──────────────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description='Amalgamate OpenCSTL headers into a single file.')
    parser.add_argument('--src-dir', default='./opencstl',
                        help='Directory containing the headers (default: ./opencstl)')
    parser.add_argument('--out', default='./opencstl.h',
                        help='Output path (default: ./opencstl.h)')
    parser.add_argument('--entry', default='opencstl.h',
                        help='Entry-point header (default: opencstl.h)')
    parser.add_argument('-c', '--compact', action='store_true',
                        help='Strip all comments and blank lines for minimum size')
    args = parser.parse_args()

    src_dir = Path(args.src_dir).resolve()
    out_path = Path(args.out).resolve()
    compact = args.compact

    if not src_dir.is_dir():
        print(f'ERROR: source directory not found: {src_dir}', file=sys.stderr)
        sys.exit(1)
    if not (src_dir / args.entry).exists():
        print(f'ERROR: entry file not found: {src_dir / args.entry}', file=sys.stderr)
        sys.exit(1)

    print(f'[amalgamate] src-dir : {src_dir}')
    print(f'[amalgamate] entry   : {args.entry}')
    print(f'[amalgamate] output  : {out_path}')
    print(f'[amalgamate] compact : {compact}')

    amalg = Amalgamator(src_dir)
    body = amalg.process_file(args.entry)
    write_output(out_path, amalg, body, compact=compact)

    # Mirror copies — skip silently if the target directory doesn't exist.
    for target in ('examples/opencstl.h',
                   'assets/opencstl.h',
                   'bench/bench_cstl/3rdparty/include/opencstl.h'):
        tgt = Path(target)
        if tgt.parent.exists():
            shutil.copy(out_path, tgt)
        else:
            print(f'[amalgamate] skip copy: {tgt.parent} does not exist',
                  file=sys.stderr)

    total = out_path.read_text(encoding='utf-8').count('\n')
    size_kb = out_path.stat().st_size / 1024
    print(f'[amalgamate] done    : {len(amalg.visited)} file(s) inlined, '
          f'{total} lines, {size_kb:.1f} KB → {out_path.name}')


if __name__ == '__main__':
    main()

File: amalgamate.pl
#!/usr/bin/env perl
# amalgamate.pl — OpenCSTL single-header amalgamator
# Usage: perl amalgamate.pl [--src-dir ./opencstl] [--out ./opencstl.h] [-c|--compact]
#
# Rules:
#   - #pragma once and include guards (#ifndef / #if !defined + #define ... #endif)
#     are stripped from each individual file and replaced with a single outer guard.
#   - #include <s> at depth-0 (unconditional) are deduplicated and hoisted to
#     the top of the output.
#   - #include <s> inside conditional blocks (#if / #ifdef / ...) are left in
#     place to preserve platform guards (e.g. Windows.h under #ifdef _WIN32).
#   - #include "local" are inlined recursively; already-visited files emit a stub.
#   - --compact strips all C/C++ comments and blank lines for minimum size.

use strict;
use warnings;
use utf8;
use open ':std', ':encoding(UTF-8)';
use Getopt::Long qw(GetOptions);
use File::Spec;
use File::Basename qw(basename dirname);
use File::Copy qw(copy);
use Cwd qw(abs_path);

# ──────────────────────────────────────────────────────────────────────────────
# Regex helpers
# ──────────────────────────────────────────────────────────────────────────────

my $RE_LOCAL_INCLUDE  = qr/^\s*#\s*include\s*"([^"]+)"/;
my $RE_SYSTEM_INCLUDE = qr/^\s*#\s*include\s*<([^>]+)>/;
my $RE_PRAGMA_ONCE    = qr/^\s*#\s*pragma\s+once\s*$/;

my $RE_GUARD_IFNDEF = qr/^\s*#\s*ifndef\s+(\w+)\s*$/;
my $RE_GUARD_IF_DEF = qr/^\s*#\s*if\s+!defined\s*\(\s*(\w+)\s*\)\s*$/;
my $RE_GUARD_DEFINE = qr/^\s*#\s*define\s+(\w+)\s*$/;

my $RE_IF_OPEN  = qr/^\s*#\s*(?:if|ifdef|ifndef)\b/;
my $RE_IF_CLOSE = qr/^\s*#\s*endif\b/;


# ──────────────────────────────────────────────────────────────────────────────
# Include-guard stripping
# ──────────────────────────────────────────────────────────────────────────────

sub strip_include_guard {
    my ($lines_ref) = @_;
    my @lines = grep { $_ !~ $RE_PRAGMA_ONCE } @$lines_ref;
    my $n = scalar @lines;

    my $idx = 0;
    $idx++ while $idx < $n && _strip($lines[$idx]) eq '';
    return \@lines if $idx >= $n;

    my $candidate_guard;
    if ($lines[$idx] =~ $RE_GUARD_IFNDEF) {
        $candidate_guard = $1;
    } elsif ($lines[$idx] =~ $RE_GUARD_IF_DEF) {
        $candidate_guard = $1;
    } else {
        return \@lines;
    }
    my $guard_open_idx = $idx;

    my $next_idx = $idx + 1;
    $next_idx++ while $next_idx < $n && _strip($lines[$next_idx]) eq '';
    return \@lines if $next_idx >= $n;

    return \@lines
        unless $lines[$next_idx] =~ $RE_GUARD_DEFINE && $1 eq $candidate_guard;

    my $guard_define_idx = $next_idx;

    my $depth = 0;
    my $closing_idx;
    for (my $j = $n - 1; $j > $guard_open_idx; $j--) {
        if ($lines[$j] =~ $RE_IF_CLOSE) {
            if ($depth == 0) {
                $closing_idx = $j;
                last;
            }
            $depth--;
        } elsif ($lines[$j] =~ $RE_IF_OPEN) {
            $depth++;
        }
    }

    return \@lines unless defined $closing_idx;

    my %skip = ($guard_open_idx => 1, $guard_define_idx => 1, $closing_idx => 1);
    my @kept;
    for (my $j = 0; $j < $n; $j++) {
        push @kept, $lines[$j] unless $skip{$j};
    }
    return \@kept;
}

# trim helper — Python's str.strip() equivalent
sub _strip {
    my ($s) = @_;
    $s =~ s/\A\s+//;
    $s =~ s/\s+\z//;
    return $s;
}


# ──────────────────────────────────────────────────────────────────────────────
# C/C++ comment stripping (for --compact mode)
# ──────────────────────────────────────────────────────────────────────────────

sub strip_c_comments {
    my ($source) = @_;
    my @out;
    my $i = 0;
    my $n = length $source;

    while ($i < $n) {
        my $c   = substr($source, $i, 1);
        my $nxt = ($i + 1 < $n) ? substr($source, $i + 1, 1) : '';

        # // line comment — skip to end of line (keep the newline)
        if ($c eq '/' && $nxt eq '/') {
            $i++ while $i < $n && substr($source, $i, 1) ne "\n";
            next;
        }

        # /* block comment */
        if ($c eq '/' && $nxt eq '*') {
            $i += 2;
            while ($i < $n - 1
                   && !(substr($source, $i, 1) eq '*'
                        && substr($source, $i + 1, 1) eq '/')) {
                $i++;
            }
            $i += 2;
            next;
        }

        # "string literal"
        if ($c eq '"') {
            push @out, $c;
            $i++;
            while ($i < $n && substr($source, $i, 1) ne '"') {
                if (substr($source, $i, 1) eq '\\' && $i + 1 < $n) {
                    push @out, substr($source, $i, 1);
                    push @out, substr($source, $i + 1, 1);
                    $i += 2;
                } else {
                    push @out, substr($source, $i, 1);
                    $i++;
                }
            }
            if ($i < $n) {
                push @out, substr($source, $i, 1);
                $i++;
            }
            next;
        }

        # 'character literal'
        if ($c eq "'") {
            push @out, $c;
            $i++;
            while ($i < $n && substr($source, $i, 1) ne "'") {
                if (substr($source, $i, 1) eq '\\' && $i + 1 < $n) {
                    push @out, substr($source, $i, 1);
                    push @out, substr($source, $i + 1, 1);
                    $i += 2;
                } else {
                    push @out, substr($source, $i, 1);
                    $i++;
                }
            }
            if ($i < $n) {
                push @out, substr($source, $i, 1);
                $i++;
            }
            next;
        }

        push @out, $c;
        $i++;
    }

    return join('', @out);
}

sub compact_source {
    my ($source) = @_;
    my $stripped = strip_c_comments($source);
    my @kept;
    for my $line (split /\n/, $stripped, -1) {
        (my $s = $line) =~ s/\s+\z//;        # rstrip
        next if _strip($s) eq '';
        push @kept, $s;
    }
    return join("\n", @kept) . "\n";
}


# ──────────────────────────────────────────────────────────────────────────────
# Core recursive processor (Amalgamator)
# ──────────────────────────────────────────────────────────────────────────────

sub new_amalgamator {
    my ($src_dir) = @_;
    return {
        src_dir          => $src_dir,
        visited          => {},   # basenames already inlined
        top_sys_includes => [],   # unconditional system includes (hoisted)
        _top_sys_seen    => {},
    };
}

sub _hoist_system_include {
    my ($amalg, $line) = @_;
    if ($line =~ $RE_SYSTEM_INCLUDE) {
        my $key = $1;
        unless ($amalg->{_top_sys_seen}{$key}) {
            $amalg->{_top_sys_seen}{$key} = 1;
            (my $rstripped = $line) =~ s/\s+\z//;
            push @{ $amalg->{top_sys_includes} }, $rstripped;
        }
    }
}

# Python-style splitlines(keepends=True): keep \n/\r\n on each line
sub _splitlines_keepends {
    my ($text) = @_;
    my @out;
    while ($text =~ /\G(.*?(?:\r\n|\n|\r|\z))/gc) {
        my $chunk = $1;
        last if $chunk eq '';
        push @out, $chunk;
    }
    return @out;
}

sub process_file {
    my ($amalg, $filename, $depth) = @_;
    $depth //= 0;

    my $path      = File::Spec->catfile($amalg->{src_dir}, $filename);
    my $canonical = basename($path);

    if ($amalg->{visited}{$canonical}) {
        return [ "// [already included: $filename]\n" ];
    }
    $amalg->{visited}{$canonical} = 1;

    unless (-e $path) {
        return [ "// [WARNING: file not found: $filename]\n" ];
    }

    open(my $fh, '<:encoding(UTF-8)', $path)
        or die "ERROR: cannot open $path: $!\n";
    my $raw_text = do { local $/; <$fh> };
    close $fh;

    my @raw = _splitlines_keepends($raw_text);
    my $stripped_ref = strip_include_guard(\@raw);

    my $sep = '=' x 78;
    my @out;
    push @out, "\n// $sep\n";
    push @out, sprintf("// BEGIN  %-30s (depth %d)\n", $filename, $depth);
    push @out, "// $sep\n\n";

    my $cond_depth = 0;   # nesting depth of conditional blocks

    for my $line (@$stripped_ref) {
        # track conditional nesting BEFORE processing the line
        if ($line =~ $RE_IF_OPEN) {
            $cond_depth++;
        } elsif ($line =~ $RE_IF_CLOSE) {
            $cond_depth--;
        }

        if ($line =~ $RE_SYSTEM_INCLUDE) {
            if ($cond_depth == 0) {
                # unconditional → hoist to top, don't emit in place
                _hoist_system_include($amalg, $line);
            } else {
                # inside #ifdef / #if → keep in place (e.g. Windows.h)
                push @out, $line;
            }
            next;
        }

        if ($line =~ $RE_LOCAL_INCLUDE) {
            my $dep = $1;
            if (-e File::Spec->catfile($amalg->{src_dir}, $dep)) {
                push @out, @{ process_file($amalg, $dep, $depth + 1) };
            } else {
                push @out, "// [external, kept as-is]\n";
                push @out, $line;
            }
            next;
        }

        push @out, $line;
    }

    push @out, "\n// $sep\n";
    push @out, "// END    $filename\n";
    push @out, "// $sep\n";

    return \@out;
}


# ──────────────────────────────────────────────────────────────────────────────
# Output assembly
# ──────────────────────────────────────────────────────────────────────────────

my $BANNER = <<'EOT';
//
// ============================================================================
//  OpenCSTL — Single-Header Amalgamation
//  Auto-generated by amalgamate.py  (do NOT edit manually)
//
//  Copyright (C) 2018-2026, Kim Bomm, all rights reserved.
//  Licensed under the OpenCSTL License Agreement.
// ============================================================================
//
EOT

my $BANNER_COMPACT = <<'EOT';
// OpenCSTL — Single-Header Amalgamation (compact)
// Copyright (C) 2018-2026, Kim Bomm. Licensed under the OpenCSTL License Agreement.
EOT

my $GUARD = 'OPENCSTL_AMALGAMATED_H';


sub write_output {
    my ($out_path, $amalg, $body_lines_ref, %opts) = @_;
    my $compact       = $opts{compact}       // 0;
    my $encoding_ansi = $opts{encoding_ansi} // 0;

    my $body_src = join('', @$body_lines_ref);
    $body_src = compact_source($body_src) if $compact;

    open(my $fh, '>:raw:encoding(UTF-8)', $out_path)
        or die "ERROR: cannot write $out_path: $!\n";

    print $fh ($compact ? $BANNER_COMPACT : $BANNER);
    print $fh "\n";
    print $fh "#pragma once\n";
    print $fh "#ifndef $GUARD\n";
    print $fh "#define $GUARD\n";
    print $fh "\n" unless $compact;

    if (@{ $amalg->{top_sys_includes} }) {
        print $fh "// ── System includes — unconditional, deduplicated ─────────────────────\n"
            unless $compact;
        for my $inc (@{ $amalg->{top_sys_includes} }) {
            print $fh $inc . "\n";
        }
        print $fh "\n" unless $compact;
    }

    print $fh $body_src;
    print $fh "\n" unless $body_src =~ /\n\z/;

    print $fh "\n" unless $compact;
    print $fh "#endif // $GUARD\n";
    close $fh;

    if ($encoding_ansi) {
        # Re-read as UTF-8 and rewrite as ASCII (errors replaced).
        open(my $rfh, '<:raw:encoding(UTF-8)', $out_path)
            or die "ERROR: cannot reopen $out_path: $!\n";
        my $text = do { local $/; <$rfh> };
        close $rfh;

        open(my $wfh, '>:raw:encoding(ASCII)', $out_path)
            or die "ERROR: cannot rewrite $out_path: $!\n";
        # Replace non-ASCII characters with '?'
        $text =~ s/([^\x00-\x7F])/?/g;
        print $wfh $text;
        close $wfh;
    }
}


# ──────────────────────────────────────────────────────────────────────────────
# Entry point
# ──────────────────────────────────────────────────────────────────────────────

sub main {
    my $src_dir_arg = './opencstl';
    my $out_arg     = './opencstl.h';
    my $entry       = 'opencstl.h';
    my $compact     = 0;
    my $ansi        = 0;
    my $help        = 0;

    GetOptions(
        'src-dir=s' => \$src_dir_arg,
        'out=s'     => \$out_arg,
        'entry=s'   => \$entry,
        'c|compact' => \$compact,
        'a|ansi'    => \$ansi,
        'h|help'    => \$help,
    ) or die "Invalid options\n";

    if ($help) {
        print <<'USAGE';
Amalgamate OpenCSTL headers into a single file.

Usage: perl amalgamate.pl [options]

Options:
  --src-dir DIR    Directory containing the headers (default: ./opencstl)
  --out FILE       Output path (default: ./opencstl.h)
  --entry FILE     Entry-point header (default: opencstl.h)
  -c, --compact    Strip all comments and blank lines for minimum size
  -a, --ansi       Use ANSI encoding
  -h, --help       Show this help message
USAGE
        return 0;
    }

    my $src_dir  = abs_path($src_dir_arg);
    my $out_path = abs_path(dirname($out_arg));
    $out_path = defined $out_path
        ? File::Spec->catfile($out_path, basename($out_arg))
        : $out_arg;

    unless (defined $src_dir && -d $src_dir) {
        print STDERR "ERROR: source directory not found: $src_dir_arg\n";
        exit 1;
    }
    unless (-e File::Spec->catfile($src_dir, $entry)) {
        print STDERR "ERROR: entry file not found: "
            . File::Spec->catfile($src_dir, $entry) . "\n";
        exit 1;
    }

    print "[amalgamate] src-dir : $src_dir\n";
    print "[amalgamate] entry   : $entry\n";
    print "[amalgamate] output  : $out_path\n";
    print "[amalgamate] compact : " . ($compact ? 'True' : 'False') . "\n";

    my $amalg = new_amalgamator($src_dir);
    my $body  = process_file($amalg, $entry);
    write_output($out_path, $amalg, $body,
                 compact => $compact, encoding_ansi => $ansi);

    # Mirror copies — skip silently if the target directory doesn't exist.
    for my $target ('examples/opencstl.h',
                    'assets/opencstl.h',
                    'bench/bench_cstl/3rdparty/include/opencstl.h',
                    'usage/opencstl.h') {
        my $tgt_dir = dirname($target);
        if (-d $tgt_dir) {
            copy($out_path, $target)
                or print STDERR "[amalgamate] copy failed: $target: $!\n";
        } else {
            print STDERR "[amalgamate] skip copy: $tgt_dir does not exist\n";
        }
    }

    open(my $rfh, '<:raw:encoding(UTF-8)', $out_path)
        or die "ERROR: cannot reopen $out_path: $!\n";
    my $final_text = do { local $/; <$rfh> };
    close $rfh;

    my $total   = () = $final_text =~ /\n/g;
    my $size_kb = (-s $out_path) / 1024;
    my $count   = scalar keys %{ $amalg->{visited} };
    printf "[amalgamate] done    : %d file(s) inlined, %d lines, %.1f KB → %s\n",
        $count, $total, $size_kb, basename($out_path);

    return 0;
}

exit main();
